Resolve #986 - HandleDisconnect function decoupling

superewald opened the pull request:
Decoupled HandleDisconnect from Game.cs into Champion.cs

Refs #986

Commit:
Resolve #986 -- HandleDisconnect function decoupling

Decoupled HandleDisconnect from game into champion class.

Refs #986

Commit:
fixed HandleDisconnect in PacketHandlerManager.cs

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/LeagueSandbox/GameServer?pullRequest=993) <br/>All committers have signed the CLA.

superewald commented on GameServerConsole/GameServerConsole.csproj:
```diff
@@ -10,7 +10,7 @@
     <AllowUnsafeBlocks>false</AllowUnsafeBlocks>
     <Optimize>false</Optimize>
-    <PlatformTarget>AnyCPU</PlatformTarget>
+    <PlatformTarget>x86</PlatformTarget>
```
This isn't regarding the issue but it resolves build settings until we have x64 libenet

Commit:
Reset changes to csproj to fix built pipeline

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/Champion.cs:
```diff
@@ -364,6 +364,23 @@ public void Respawn()
+                    _game.PacketNotifier.NotifyUnitAnnounceEvent(UnitAnnounces.SUMMONER_DISCONNECTED, peerinfo.Champion);
+                }
+                peerinfo.IsDisconnected = true;
+                peerinfo.Champion.StopChampionMovement();
```
Change `peerinfo.Champion` to `this` in all occurrences.  Also, do not use `peerInfo` and so here. Try to put all these things in `Game` class, to decouple Peer from the logic of the objects.

Also, do you know why `HandleDisconnect` should check if the peer is disconnected, does it really get called when the peer is not disconnected?

superewald replied:
I forgot to refactor that function, kill me. 

I thought that check maybe was to prevent double calling the packetnotifier, but I tested without and it worked fine so I removed it.

danil179 commented on GameServerCore/Domain/GameObjects/IChampion.cs:
```diff
@@ -23,6 +23,7 @@ public interface IChampion : IObjAiBase
         bool CanCast();
         void Recall();
         void Respawn();
+        bool HandleDisconnect();
```
Maybe rename to `OnDisconnect` to comply with the general style of C#.

danil179 reviewed (changes requested):
Need to decouple peer from Champion class.

Commit:
- moved peer logic to packethandlermanager.cs
- added necessary IPacketNotifier to IPacketHandlerManager
- renamed disconnect event handler in Champion.cs
- removed unnecessary code

danil179 commented on PacketDefinitions420/PacketHandlerManager.cs:
```diff
@@ -244,8 +244,16 @@ public bool HandlePacket(Peer peer, byte[] data, Channel channelId)
+
+            if (peerInfo != null)
+            {
+                _game.PacketNotifier.NotifyUnitAnnounceEvent(UnitAnnounces.SUMMONER_DISCONNECTED, peerInfo.Champion);
```
Future suggestion: Maybe we want to take this line inside champion at some point. Currently, it is not possible as `peerInfo==null` is possible with disconnect after disconnect maybe???

danil179 reviewed (approved):
Approved.
